File: src/main/java/org/blitzortung/android/app/preference/PreferenceKey.java
package org.blitzortung.android.app.preference;

import android.content.SharedPreferences;
import org.blitzortung.android.app.controller.LocationHandler;
import org.blitzortung.android.data.provider.DataProviderType;

import java.util.HashMap;
import java.util.Map;

public enum PreferenceKey {
    USERNAME("username", ""),
    PASSWORD("password", ""),
    RASTER_SIZE("raster_size", "10000"),
    MAP_TYPE("map_mode", "SATELLITE"),
    MAP_FADE("map_fade"),
    COLOR_SCHEME("color_scheme"),
    QUERY_PERIOD("query_period"),
    BACKGROUND_QUERY_PERIOD("background_query_period"),
    SHOW_PARTICIPANTS("show_participants"),
    SHOW_LOCATION("location"),
    ALARM_ENABLED("alarm_enabled"),
    ALARM_SOUND_SIGNAL("alarm_sound_signal"),
    ALARM_VIBRATION_SIGNAL("alarm_vibration_signal"),
    NOTIFICATION_DISTANCE_LIMIT("notification_distance_limit"),
    SIGNALING_DISTANCE_LIMIT("signaling_distance_limit"),
    REGION("region", "1"),
    DATA_SOURCE("data_source", DataProviderType.HTTP.toString()),
    MEASUREMENT_UNIT("measurement_unit"),
    DO_NOT_SLEEP("do_not_sleep"),
    INTERVAL_DURATION("interval_duration", "120"),
    HISTORIC_TIMESTEP("historic_timestep", "30"),
    LOCATION_MODE("location_mode", LocationHandler.LocationProvider.NETWORK.getType()),
    LOCATION_LONGITUDE("location_longitude", "11.0"),
    LOCATION_LATITUDE("location_latitude", "49.0");
    
    private final String key;
    private final String defaultValue;

    private PreferenceKey(String key, String defaultValue) {
        this.key = key;
        this.defaultValue = defaultValue;
    }

    @Override
    public String toString()
    {
        return key;
    }

    private static final Map<String, PreferenceKey> stringToValueMap = new HashMap<String, PreferenceKey>();
    static {
        for (PreferenceKey key : PreferenceKey.values()) {
            String keyString = key.toString();
            if (stringToValueMap.containsKey(keyString)) {
                throw new IllegalStateException(String.format("key value '%s' already defined", keyString));
            }
            stringToValueMap.put(keyString, key);
        }
    }

    public static PreferenceKey fromString(String string) {
        return stringToValueMap.get(string);
    }

    public String getValue(SharedPreferences sharedPreferences) {
        return sharedPreferences.getString(key, defaultValue);
    }
}

File: common/src/main/java/org/blitzortung/android/common/preferences/PreferenceKey.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.common.preferences

enum class PreferenceKey(val key: String) {
    USERNAME("username"),
    PASSWORD("password"),
    SERVICE_URL("service_url"),
    RASTER_SIZE("raster_size"),
    COUNT_THRESHOLD("count_threshold"),
    MAP_TYPE("map_mode"),
    MAP_FADE("map_fade"),
    COLOR_SCHEME("color_scheme"),
    QUERY_PERIOD("query_period"),
    BACKGROUND_QUERY_PERIOD("background_query_period"),
    SHOW_PARTICIPANTS("show_participants"),
    SHOW_LOCATION("location"),
    ALERT_ENABLED("alarm_enabled"),
    ALERT_SOUND_SIGNAL("alarm_sound_signal"),
    ALERT_VIBRATION_SIGNAL("alarm_vibration_signal"),
    ALERT_NOTIFICATION_DISTANCE_LIMIT("notification_distance_limit"),
    ALERT_SIGNALING_DISTANCE_LIMIT("signaling_distance_limit"),
    ALERT_SIGNALING_THRESHOLD_TIME("signaling_threshold_time"),
    REGION("region"),
    DATA_SOURCE("data_source"),
    MEASUREMENT_UNIT("measurement_unit"),
    DO_NOT_SLEEP("do_not_sleep"),
    INTERVAL_DURATION("interval_duration"),
    HISTORIC_TIMESTEP("historic_timestep"),
    LOCATION_MODE("location_mode"),
    LOCATION_LONGITUDE("location_longitude"),
    LOCATION_LATITUDE("location_latitude");

    override fun toString(): String {
        return key
    }

    companion object {

        private val stringToValueMap = java.util.HashMap<String, PreferenceKey>()

        init {
            for (key in org.blitzortung.android.common.preferences.PreferenceKey.values()) {
                val keyString = key.toString()
                if (keyString in org.blitzortung.android.common.preferences.PreferenceKey.Companion.stringToValueMap) {
                    throw IllegalStateException("key value '%s' already defined".format(keyString))
                }
                org.blitzortung.android.common.preferences.PreferenceKey.Companion.stringToValueMap[keyString] = key
            }
        }

        fun fromString(string: String): org.blitzortung.android.common.preferences.PreferenceKey {
            return org.blitzortung.android.common.preferences.PreferenceKey.Companion.stringToValueMap[string]!!
        }
    }
}

interface OnSharedPreferenceChangeListener : android.content.SharedPreferences.OnSharedPreferenceChangeListener {
    fun onSharedPreferencesChanged(sharedPreferences: android.content.SharedPreferences, vararg keys: org.blitzortung.android.common.preferences.PreferenceKey) {
        keys.forEach { onSharedPreferenceChanged(sharedPreferences, it) }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: android.content.SharedPreferences, keyString: String) {
        onSharedPreferenceChanged(sharedPreferences, org.blitzortung.android.common.preferences.PreferenceKey.Companion.fromString(keyString))
    }

    fun onSharedPreferenceChanged(sharedPreferences: android.content.SharedPreferences, key: org.blitzortung.android.common.preferences.PreferenceKey)
}
